Fix #411 - Add markers for boat, train, subway, and tram

bendu opened the pull request:
Icons taken from https://github.com/google/material-design-icons/tree/master/maps/svg/production

This is to start addressing #411 I was going to make the subway icon more like the tram and the train icon, but I could not figure out how to make it happen in inkscape and my knowledge of SVG is limited. If these icons are acceptable, I will make the various arrow direction icons. I already have some code ready to support displaying different vehicle icons, which I will add in a future PR.

Also, how do I generate the png files from svg?

barbeau commented:
Thanks @bendu!  I don't currently have access to my machine so I can't look at the SVGs now on mobile (AFAIK), but I'll take a look early next week.

IIRC I used the Android Icon Generator and just dragged the SVGs onto the page to generate the PNGs:
https://romannurik.github.io/AndroidAssetStudio/icons-generic.html#source.space.trim=1&source.space.pad=0&size=24&padding=0&color=000%2C100&name=ic_example

barbeau commented:
Also, I'd prefer a single PR with images and code together for final merge, although this is fine for a WIP.

barbeau commented:
@bendu Finally looking at this on my machine.

Ferry/boat icon looks fine:

![image](https://cloud.githubusercontent.com/assets/928045/17488450/7f255de2-5d67-11e6-8f7d-b37439afdd9b.png)

Subway needs some work - here's what's currently in the PR:

![image](https://cloud.githubusercontent.com/assets/928045/17488506/b5c14ac8-5d67-11e6-9b6c-1534b195cf95.png)

The white/black portion should be inverted so the vehicle is white.  Also, I'd prefer using the material design subway icon (white/black needs to be inverted here too):
https://github.com/google/material-design-icons/blob/master/maps/svg/production/ic_directions_subway_48px.svg

![image](https://cloud.githubusercontent.com/assets/928045/17488578/fccf1d50-5d67-11e6-939d-f963245f5cad.png)

It's been a while since I did the originals in Inkscape, and I don't use it regularly - I'll see if I can take a look to remember how I did the inversion there.

Train looks good:

![image](https://cloud.githubusercontent.com/assets/928045/17488634/38273bf8-5d68-11e6-8d7a-496d6fcf15ea.png)

Tram looks good:

![image](https://cloud.githubusercontent.com/assets/928045/17488659/66b3e192-5d68-11e6-8ac5-4664e1b678df.png)

bendu commented:
@barbeau I have updated the PR with code for showing different vehicles in VehicleOverlay. I also updated the subway icon. I added markers for the various directions as well. Once the icons look good. I will generate the png files. There are 36 svg files to generate, and it looks like I will have to do them 1 by 1, so I am waiting until these icons are OK'd before proceeding.

![subway](https://cloud.githubusercontent.com/assets/5668844/17654143/8c2d0444-6254-11e6-97eb-c6abbaf4ebdc.PNG)

bendu commented:
@barbeau I updated the commit with png files of the 4 markers. Here are some screenshots of a few non-bus vehicles I could find.

![ferry](https://cloud.githubusercontent.com/assets/5668844/17724984/bbcef56a-63fb-11e6-9736-52e9af584e7e.PNG)

![link](https://cloud.githubusercontent.com/assets/5668844/17724985/bbd0bfe4-63fb-11e6-894f-d95df753f6ec.PNG)

barbeau commented:
@bendu This looks good!  All icons look great so far.

The GTFS spec defines the following [route types](https://developers.google.com/transit/gtfs/reference?hl=en#routestxt) - here's what I believe we have covered so far (including master branch and this PR):

- [x] 0 - Tram, Streetcar, Light rail. Any light rail or street level system within a metropolitan area - *marker_with_tram_smaller_all_inside.svg*
- [x] 1 - Subway, Metro. Any underground rail system within a metropolitan area - *marker_with_subway_smaller_all_inside.svg*
- [ ] 2 - Rail. Used for intercity or long-distance travel - 
- [x] 3 - Bus. Used for short- and long-distance bus routes - *marker_with_bus_smaller_all_inside.svg*
- [x] 4 - Ferry. Used for short- and long-distance boat service - *marker_with_boat_smaller_all_inside.svg*
- [x] 5 - Cable car. Used for street-level cable cars where the cable runs beneath the car - *marker_with_tram_smaller_all_inside.svg*
- [ ] 6 - Gondola, Suspended cable car. Typically used for aerial cable cars where the car is suspended from the cable
- [ ] 7 - Funicular. Any rail system designed for steep inclines

I believe "2) Rail" could use:
https://github.com/google/material-design-icons/blob/master/maps/svg/production/ic_directions_railway_48px.svg

I believe "5) Cable car" can just use the "Tram" icon, as indicated above.

I'm fine omitting "Gondola" and "Funicular" for now :).

bendu commented:
@barbeau I am using marker_with_train_smaller_all_inside.svg icon for rail (2). 
I will update "cable car" to use the "tram" icon.

barbeau commented:
> I am using marker_with_train_smaller_all_inside.svg icon for rail (2).

Whoops!  Sorry, somehow I missed that.  Here are the two side by side for reference - train on left, railway on right:

![image](https://cloud.githubusercontent.com/assets/928045/17760273/9d4e78c8-64cc-11e6-903b-26cc80553aa7.png)

I'm fine with train icon for 2) Rail - that icon looks more like what I would think of for rail.  railway icon is easier to confuse with subway as well.

bendu commented:
Here is a short script I used for generating the various icons with the direction arrows (written in powershell) https://gist.github.com/bendu/d87a66d4c014e0fca9fbad4a9fe94c18

barbeau commented:
@bendu That's awesome, thanks for sharing!   Definitely much faster than editing by hand.  Would you mind adding this to the same folder as the SVGs in the repo and add a README in markdown that describes the script and SVGs?

bendu commented:
Updated the code to use tram icon for cable car, added the svg script.
On Aug 20, 2016 10:31 AM, Sean Barbeau <notifications@github.com> wrote:@bendu That's awesome, thanks for sharing!   Definitely much faster than editing by hand.  Would you mind adding this to the same folder as the SVGs in the repo and add a README in markdown that describes the script and SVGs?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

barbeau commented:
Thanks!  Is this ready for final review, or are you still working on it?  Either is fine, just don't want to drop the ball.

bendu commented:
Ready for final review
On Aug 21, 2016 2:33 PM, Sean Barbeau <notifications@github.com> wrote:Thanks!  Is this ready for final review, or are you still working on it?  Either is fine, just don't want to drop the ball.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

barbeau commented on onebusaway-android/src/amazon/java/org/onebusaway/android/map/googlemapsv2/VehicleOverlay.java:
```diff
@@ -91,25 +92,30 @@
 
     private static final int NORTH = 0;
 
-    private static final int NORTH_WEST = 1;
```
Have you tested to make sure that these changes in numeric indexes don't result in flipped icons?  For example, `2` used to be `WEST` and now it's `EAST`.  IIRC the halfWinds generated by `MathUtils` go in counter-clockwise order, which is why this was reversed.  If that's correct, then from a quick scan of the code I believe the "WEST" direction would now result in an "EAST" icon?

If this is correct, then we definitely need better comments (i.e., I should have added better comments), as this wasn't clear in the code itself.

bendu replied:
IIRC the const direction vars were only used to index into the vehicle_icons array. I redefined the numeric indexes to be consistent with the halfWinds generated. I will double check when I get the chance.

barbeau replied:
I'll test more myself too.  I loaded it up and at a glance icons looked correct based on route directions.

bendu replied:
I tested the app and the icons are pointing the right direction. I looked over the diff and I found that the halfwinds do go clockwise. The constants were used in the former method createvehicleicon to retrieve the icon for a particular direction. I now use the constants to determine which index the icon should go, which was not the case before.

barbeau commented on onebusaway-android/src/amazon/java/org/onebusaway/android/map/googlemapsv2/VehicleOverlay.java:
```diff
@@ -91,25 +92,30 @@
+    // Vehicle type (if available) -> icon set
+    private static Map<Integer, Bitmap[]> vehicle_type_icons = new HashMap<>(ObaRoute.NUM_TYPES);
+
+//    private static Bitmap[] vehicle_icons = new Bitmap[NUM_DIRECTIONS];  // Black vehicle icons
```
Looks like this is left over from the previous implementation, you can just remove it.

bendu replied:
will do

barbeau commented on onebusaway-android/src/amazon/java/org/onebusaway/android/map/googlemapsv2/VehicleOverlay.java:
```diff
@@ -182,21 +188,11 @@ public synchronized void clear() {
-            vehicle_icons[7] = createVehicleIcon(NORTH_WEST);
-            vehicle_icons[8] = createVehicleIcon(NO_DIRECTION);
-        }
+        vehicle_type_icons.put(ObaRoute.TYPE_BUS, createBusIcons());
```
IIRC because `vehicle_type_icons` is static, it can end up being initialized multiple times, each time `VehicleOverlay` is instantiated.  I had the check for `vehicle_icons[0] == null` in the previous code to make sure that we didn't pay this initialization penalty more than once.  This is even more important now that we're loading more vehicle icon types.  You can add a similar IF block around the initialization of `vehicle_type_icons`, something like `if (!vehicle_type_icons.contains(ObaRoute.TYPE_BUS)) {`, to do this same type of check.

bendu replied:
good point.

bendu commented:
@barbeau I've address your code review comments and updated the diff.

barbeau commented:
Thanks @bendu, this looks great!  No more floating buses :+1:  :)

barbeau commented:
We may want to further optimize icon loading (it's doubtful that in any one execution all 5 vehicle types will be displayed), but that can happen in another issue.

barbeau commented:
Opened a new issue for optimizing black template icon loading at https://github.com/OneBusAway/onebusaway-android/issues/687.
